### PR TITLE
Introducing a new pass to distribute experts

### PR DESCRIPTION
**Pull request by shaahji**
Introducing a new pass to distribute experts in onnx model by splitting the graph making it capable of running inferencing across multiple GPUs.

### CODE REVIEW

**Review by devang-ml — CHANGES_REQUESTED** [review]

File: olive/passes/onnx/distribute_experts.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        slice_nodes = [n for n in consumers[reshape_node_oname] if n.op_type == 'Slice']
+
+        concat_node = nodes[expert[13]]
+        matmul_nodes = [producers[iname] for iname in concat_node.input]
+
+        num_experts = len(consumers[div_node_oname])
+        experts_per_rank = int(num_experts / world_size)
+
+        for i in range(experts_per_rank, num_experts):
+            mul_nodes[i].input.remove(div_node_oname)
```
Comment (devang-ml) [inline olive/passes/onnx/distribute_experts.py:86]: If there are 16 experts and world size is 4 then does this mean all ranks will get first 4 experts?

Comment (devang-ml) [reply olive/passes/onnx/distribute_experts.py:86]: Each rank should get different set of 4 experts.

**Review by devang-ml — CHANGES_REQUESTED** [review]

**Comment by xiaoyu-work** [comment]
A few comments:
1. Is this only for GPU? Can you please add a check here if this pass was run with other devices like CPU?
2. Can you please add a unit test for this pass? You can reference the existing pass unit tests.
3. You can run `pre-commit run --all-files` before commit to reformat your code.
4. Can you also update the document for this new pass?

File: olive/passes/onnx/distribute_experts.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        }
+
+
+    def _run_for_config(self, model: ONNXModel, config: Dict[str, Any], output_model_path: str) -> ONNXModel:
+        distribute_experts(config['world_size'], model.model_path, output_model_path)
+        return ONNXModel(output_model_path, model.name)
+
+
+def _main():
+    parser = argparse.ArgumentParser()
```
[the remark this answers is not in the document]
Comment (shaahji) [reply olive/passes/onnx/distribute_experts.py:214]: This allows running the script in standalone mode. I will move it out into a local script.

File: olive/passes/onnx/distribute_experts.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+                    "Number of GPU nodes to distribute the model for. Must be greater than 1."
+                ),
+            ),
+        }
+
+
+    def _run_for_config(self, model: ONNXModel, config: Dict[str, Any], output_model_path: str) -> ONNXModel:
+        distribute_experts(config['world_size'], model.model_path, output_model_path)
+        return ONNXModel(output_model_path, model.name)
+
```
[the remark this answers is not in the document]
Comment (shaahji) [reply olive/passes/onnx/distribute_experts.py:211]: Yes, but that doesn't exist yet in the branch. Will update it once the other branch is merged.

**Pull request closed without merging**